### PR TITLE
feat(extra-natives/five): GET_VEHICLE_GEAR_RATIO and SET_VEHICLE_GEAR_RATIO 

### PR DESCRIPTION
**Pull request by RickyB505**
### Goal of this PR

These new natives will allow developers to create scripts to create custom gear ratios for vehicles.

### How is this PR achieving the goal

I added the memory editing stuff to VehicleExtraNatives.cpp to allow me to create the 2 new natives.

### The natives
float GetVehicleGearRatio(Vehicle vehicle, int gear);
void SetVehicleGearRatio(Vehicle vehicle, int gear, float ratio);

### This PR applies to the following area(s)

FiveM, Vehicle, Lua, JS, C#

### Successfully tested on

**Game builds:**  3095,2944,2802,2699,2612,2545,2372,2189,2060,1604

**Platforms:** Windows, Linux

### Checklist

- [x] Code compiles and has been tested successfully.
- [x] Code explains itself well and/or is documented.
- [x] My commit message explains what the changes do and what they are for.
- [x] No extra compilation warnings are added by these changes.

### CODE REVIEW

**Comment by thorium-cfx** [comment]
Hey Ricky,

Thank you for the PR, seems like you got it working, good job! Some prechecks: your commit titles don't follow our format, see https://github.com/citizenfx/fivem/actions/runs/8118915693/job/22194013573?pr=2409#step:4:83 could you resolve these?

When you are at it can you also squash the commits? Guide: https://docs.fivem.net/docs/contributing/git/squash-guide/

**Comment by RickyB505** [comment]
> Hey Ricky,
> 
> Thank you for the PR, seems like you got it working, good job! Some prechecks: your commit titles don't follow our format, see https://github.com/citizenfx/fivem/actions/runs/8118915693/job/22194013573?pr=2409#step:4:83 could you resolve these?
> 
> When you are at it can you also squash the commits? Guide: https://docs.fivem.net/docs/contributing/git/squash-guide/

should be all good now just need to let the thing run

File: code/components/extra-natives-five/src/VehicleExtraNatives.cpp [hunk 1]
```diff
@@ -866,6 +878,24 @@ static HookFunction initFunction([]()
 		};
 	};
 
+	fx::ScriptEngine::RegisterNativeHandler("GET_VEHICLE_GEAR_RATIO", [](fx::ScriptContext& context)
+	{
+		unsigned char gear = context.GetArgument<int>(1);
+		if (fwEntity* vehicle = getAndCheckVehicle(context, "GET_VEHICLE_GEAR_RATIO"))
+		{
+			context.SetResult<float>(*(float*)((char*)vehicle + VehicleGearRatioOffset + gear * sizeof(float)));
```
Comment (Disquse) [inline code/components/extra-natives-five/src/VehicleExtraNatives.cpp:886]: You must ensure the user won't be able to read or write out of bounds. As of b3095, it seems the biggest possible amount of gears is `<= 10`. You need to check that the passed `gear` argument isn't bigger than 10. You can also use `*(uint8_t*)(CTransmission+6)` to get this value directly from memory.
```
// The same code block we're used to calculate "VehicleGearRatioOffset".
VehicleMaxGearOffset = CurrentGearOffset + 6;
```

Here's a pattern to the function where this max gear value is being set: `BA ? ? ? ? F3 0F 11 49`

File: code/components/extra-natives-five/src/VehicleExtraNatives.cpp [hunk 2]
```diff
@@ -866,6 +878,24 @@ static HookFunction initFunction([]()
...
+		}
+	});
+
+	fx::ScriptEngine::RegisterNativeHandler("SET_VEHICLE_GEAR_RATIO", [](fx::ScriptContext& context)
+	{
+		unsigned char gear = context.GetArgument<int>(1);
+		if (fwEntity* vehicle = getAndCheckVehicle(context, "SET_VEHICLE_GEAR_RATIO"))
+		{
+			*(float*)((char*)vehicle + VehicleGearRatioOffset + gear * sizeof(float)) = context.GetArgument<float>(2);
```
Comment (Disquse) [inline code/components/extra-natives-five/src/VehicleExtraNatives.cpp:895]: Same as above.

File: ext/native-decls/GetVehicleGearRatio.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+---
+ns: CFX
+apiset: client
+---
```
Comment (Disquse) [inline ext/native-decls/GetVehicleGearRatio.md:4]: Please add `game: gta5` here on the next line after 'apiset'.

File: ext/native-decls/GetVehicleGearRatio.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+## GET_VEHICLE_GEAR_RATIO
+
+```c
+float GET_VEHICLE_GEAR_RATIO(Vehicle vehicle, int gear);
+```
+Gets vehicles gear ratio on choosen gear.
```
Comment (Disquse) [inline ext/native-decls/GetVehicleGearRatio.md:10]: Insert a newline after the end of the code block.

File: ext/native-decls/GetVehicleGearRatio.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+## Examples
+```lua
+local Vehicle = GetVehiclePedIsIn(PlayerPedId(-1))
+local currentGear = GetVehicleCurrentGear(Vehicle)
+
+print(GetVehicleGearRatio(Vehicle, currentGear)) -- will print current vehicle gear to console
+```
+
+```js
+const Vehicle = GetVehiclePedIsIn(PlayerPedId(-1));
```
Comment (Disquse) [inline ext/native-decls/GetVehicleGearRatio.md:25]: Please maintain consistent usage of the same codestyle throughout, if `lowerCamelCase` is used for variables in JS examples, keep using it across all JS examples.

File: ext/native-decls/SetVehicleGearRatio.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
+---
+ns: CFX
+apiset: client
+---
+## SET_VEHICLE_GEAR_RATIO
+
+```c
+void SET_VEHICLE_GEAR_RATIO(Vehicle vehicle, int gear, float ratio);
+```
+Sets the vehicles gear ratio on choosen gear, reverse gear needs to be a negative float and forward moving gear needs to be a positive float. Refer to the examples if confused.
```
Comment (Disquse) [inline ext/native-decls/SetVehicleGearRatio.md:10]: Insert a newline after the end of the code block.

File: ext/native-decls/SetVehicleGearRatio.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+## Parameters
```
Comment (Disquse) [inline ext/native-decls/SetVehicleGearRatio.md:11]: Insert a newline after the end of the description.

File: ext/native-decls/SetVehicleGearRatio.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    SetVehicleGearRatio(Vehicle, 0, -3.32)  -- reverse gear at -3.21:1
+    SetVehicleGearRatio(Vehicle, 1, 4.71)   -- 1st gear at 4.71:1
+    SetVehicleGearRatio(Vehicle, 2, 3.14)   -- 2nd gear at 3.14:1
+    SetVehicleGearRatio(Vehicle, 3, 2.11)   -- 3rd gear at 2.11:1
+    SetVehicleGearRatio(Vehicle, 4, 1.67)   -- 4th gear at 1.67:1
+    SetVehicleGearRatio(Vehicle, 5, 1.29)   -- 5th gear at 1.29:1
+    SetVehicleGearRatio(Vehicle, 6, 1.0)    -- 6th gear at 1.0:1
+    SetVehicleGearRatio(Vehicle, 7, 0.84)   -- 7th gear at 0.84:1
+    SetVehicleGearRatio(Vehicle, 8, 0.67)   -- 8th gear at 0.67:1
+    return
```
Comment (Disquse) [inline ext/native-decls/SetVehicleGearRatio.md:28]: "return" isn't necessary here and in other examples too.

**Review by Disquse — CHANGES_REQUESTED** [review]

**Comment by gottfriedleibniz** [comment]
Should we handle or document [MODIFY_VEHICLE_TOP_SPEED](https://docs.fivem.net/natives/?_0x93A3996368C94158), [SET_VEHICLE_MOD](https://docs.fivem.net/natives/?_0x6AF0636DDEDCB6DD), [REMOVE_VEHICLE_MOD](https://docs.fivem.net/natives/?_0x92D619E420858204) (VMT_GEARBOX) potentially recalculating this gear ratio array and overwriting changes made by `SET_VEHICLE_GEAR_RATIO`? Additionally b3095 introduced `0x337EF33DA3DDB990` which also repopulates the array while allowing different constants, e.g., `0.9 vs. 0.8` and `1.1 vs 1.03`

**Comment by RickyB505** [comment]
> Should we handle or document [MODIFY_VEHICLE_TOP_SPEED](https://docs.fivem.net/natives/?_0x93A3996368C94158), [SET_VEHICLE_MOD](https://docs.fivem.net/natives/?_0x6AF0636DDEDCB6DD), [REMOVE_VEHICLE_MOD](https://docs.fivem.net/natives/?_0x92D619E420858204) (VMT_GEARBOX) potentially recalculating this gear ratio array and overwriting changes made by `SET_VEHICLE_GEAR_RATIO`? Additionally b3095 introduced `0x337EF33DA3DDB990` which also repopulates the array while allowing different constants, e.g., `0.9 vs. 0.8` and `1.1 vs 1.03` in its calculations (native is tentatively named `_SET_TRANSMISSION_REDUCED_GEAR_RATIO`).

possibly might want to add to the corresponding decals that they might interfere with each other

**Comment by RickyB505** [comment]
is there anything else i need to change about this?

File: code/components/extra-natives-five/src/VehicleExtraNatives.cpp [hunk 9]
```diff
@@ -866,6 +869,40 @@ static HookFunction initFunction([]()
...
+		if (fwEntity* vehicle = getAndCheckVehicle(context, "GET_VEHICLE_GEAR_RATIO"))
+		{
+			if (gear <= 10)
+			{
+				context.SetResult<float>(*(float*)((char*)vehicle + VehicleGearRatioOffset + gear * sizeof(float)));
+			}
+			else
+			{
+				context.SetResult<float>(0.0f);
+				return;
```
Comment (Disquse) [inline code/components/extra-natives-five/src/VehicleExtraNatives.cpp:884]: You don't need a return here.

File: code/components/extra-natives-five/src/VehicleExtraNatives.cpp [hunk 10]
```diff
@@ -866,6 +869,40 @@ static HookFunction initFunction([]()
...
+	fx::ScriptEngine::RegisterNativeHandler("SET_VEHICLE_GEAR_RATIO", [](fx::ScriptContext& context)
+	{
+		unsigned char gear = context.GetArgument<int>(1);
+		if (fwEntity* vehicle = getAndCheckVehicle(context, "SET_VEHICLE_GEAR_RATIO"))
+		{
+			if (gear <= 10)
+			{
+				*(float*)((char*)vehicle + VehicleGearRatioOffset + gear * sizeof(float)) = context.GetArgument<float>(2);
+			}
+			else
```
Comment (Disquse) [inline code/components/extra-natives-five/src/VehicleExtraNatives.cpp:898]: You don't need to set result for a `SET_` native, so this whole `else` block is useless.

File: ext/native-decls/SetVehicleGearRatio.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+Sets the vehicles gear ratio on choosen gear, reverse gear needs to be a negative float and forward moving gear needs to be a positive float. Refer to the examples if confused.
+
+## Parameters
+* **vehicle**: The vehicle handle.
+* **gear**: The vehicles gear you want to change.
+* **ratio**: The gear ratio you want to use.
+
+## Examples
+```lua
```
Comment (Disquse) [inline ext/native-decls/SetVehicleGearRatio.md:20]: Place a newline after `## Examples`.

File: ext/native-decls/GetVehicleGearRatio.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+```
+
+Gets vehicles gear ratio on choosen gear.
+
+## Parameters
+* **vehicle**: The vehicle handle.
+* **gear**: The vehicles gear you want to get.
+
+## Examples
+```lua
```
Comment (Disquse) [inline ext/native-decls/GetVehicleGearRatio.md:19]: Place a newline after `## Examples`.

**Review by Disquse — CHANGES_REQUESTED** [review]
A few more small changes and it's good to go!

**Comment by RickyB505** [comment]
@Disquse there you go all done lmk if you notice anything else that needs changing

**Comment by RickyB505** [comment]
any updates for changes needed?

**Comment by RickyB505** [comment]
i assume this is ready to merge then?

**Comment by Disquse** [comment]
Tested this code in-game, works great. Thanks for your contribution; this is a nice feature to have. Apologies for the very slow review process; I was busy with a lot of other stuff. I'm attaching a debug script I made to test this feature, just in case anyone is interested in testing it out before there are any proper resources: https://pastebin.com/yDYj0sqA

**Review by Disquse — APPROVED** [review]

**Pull request merged**